### DISCUSSION TITLE
Allow Guid as an attribute parameter type

### DISCUSSION THREAD

**Discussion by jpritchard-isl** [dotnet/csharplang#7090] [Language Ideas]
To enable the passing use of Guids as parameters on a customer attribute, add Guid to the attribute parameter types.

Currently it is necessary and typical to workaround this limitation by passing string forms of Guids.
```csharp
[CustomAttribute("{B54472B8-CB74-40E4-A6B1-52771502D273}")]
```
where the Guids are either hard coded or defined as constants.  Either the developer must maintain a string and a Guid constant of the same value or explicitly `Parse(...)` / `ToString()` throughout their code.

Example usage
```csharp
using System;

public static class Flavours
{
    public const Guid APPLE = new("{83E4E771-9DFA-42C7-89D9-51979E856417}");
    public const Guid ORANGE = new("{14E1D45F-C0C9-41C4-A772-2C4DE79B8DD1}");
    public const Guid MELON = new("{B6085AD4-6F24-4426-8E1E-E12935648B52}");
    public const Guid LEMON = new("{654680C3-BCFA-4527-9483-8B158977F453}");
}

public class RequiredFruitsAttribute : Attribute
{
    public RequiredFruitsAttribute(params Guid[] requiredFruits)
    {
        ...
    }
}

public class MyMachine
{
    [RequiredFruits(Flavours.APPLE, Flavours.MELON)]
    public void MyMethod()
    {

    }
}
```

There have been proposals for other types similarly be added such as Decimal.  It is probable that allowing new types will require a CLR enhancement, and therefore it would be prudent to consider enabling the use of any type on an attribute parameter.

**Comment by HaloFour** [comment]
The types supported in attributes is limited in the runtime, as it needs to understand how to decode the binary metadata back into the type at runtime when those attributes are read through reflection.

IMO, Guid is not a particularly compelling case for expanding the list of types.  The approach of using a string has been known and widely used since C# 1.0, and given Guids were much more commonly used then for scenarios like COM interop, it doesn't seem that the runtime or language teams felt that using a string to represent them was onerous enough.  I could see Guid coming along for the ride if blittable types were supported in general, but probably not otherwise.

Although adding the type to the runtime support would only be half of what would need to happen.  The other problem is that you'd also need the compiler to understand how to interpret any of the expressions to parse a Guid into a compile-time constant.  This is also why Guids can't be used as `const` today.
